### PR TITLE
Plate classifications

### PR DESCRIPTION
**Pull request by MattsonCam**
In this pr plates 1 and 2 were trained and evaluated with logistic regression models, and the previous combined plate 3 and 3 prime logistic regression code was refactored. Future training and evaluation of models will likely follow a similar structure. If you have any questions or suggestions please feel free to let me know!

### CODE REVIEW

File: 0.data_analysis/plate_projection_cp_fs_data/nbconverted/plate_1_2_projections.py [hunk 1]
```diff
@@ -0,0 +1,195 @@
+#!/usr/bin/env python
+# coding: utf-8
+
+# # Create projections with PCA and UMAP
+
+# ## Imports
+
+# In[ ]:
```
Comment (gwaybio) [inline 0.data_analysis/plate_projection_cp_fs_data/nbconverted/plate_1_2_projections.py:8]: these are appearing as non-executed code blocks. Please make sure to add sequentially executed notebooks

File: 1.train_models/log_reg_plate_1_cp_fs_data/nbconverted/train_model.py [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+# In[ ]:
+
+
+target_column = "Metadata_genotype"
+stratify_column = "Metadata_Well"
+
+# These represent the fractions of the entire dataset
+train_val_frac = 0.85
+test_frac = 1 - train_val_frac
+val_frac = 0.15
```
Comment (gwaybio) [inline 1.train_models/log_reg_plate_1_cp_fs_data/nbconverted/train_model.py]: Is this expected to add to 1? (train + test + val)

File: 1.train_models/log_reg_plate_1_cp_fs_data/nbconverted/train_model.py [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+# # Model Training
+
+# In[ ]:
+
+
+# Default number of splits
+num_splits = 5
+
+
+def kcross_val(model, X_data, y_data, splits=num_splits):
```
Comment (gwaybio) [inline 1.train_models/log_reg_plate_1_cp_fs_data/nbconverted/train_model.py]: Maybe worth adding this function to a separate `.py` file that you can import?

Comment (gwaybio) [reply 1.train_models/log_reg_plate_1_cp_fs_data/nbconverted/train_model.py]: Also, just checking, but do you need to implement this logic? Doesn't sklearn contain functions to assist with this? I see `KFold` below, but what about [`cross_val_score`](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.cross_val_score.html#sklearn.model_selection.cross_val_score)?

[`cross_validate`](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.cross_validate.html#sklearn.model_selection.cross_validate) also comes to mind

Whenever possible, we should not reinvent wheels :)

Comment (gwaybio) [reply 1.train_models/log_reg_plate_1_cp_fs_data/nbconverted/train_model.py]: My last comment about this, is that I'm not sure we should be selecting models based on accuracy. I think average precision is probably a better metric - here are others that can work in the predefined sklearn functions: https://scikit-learn.org/stable/modules/model_evaluation.html#scoring-parameter

File: 1.train_models/log_reg_plate_1_cp_fs_data/nbconverted/train_model.py [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+kcross_val(lr, X_data=traindf.drop("label", axis="columns"), y_data=traindf["label"])
+
+
+# ## Save Data
+
+# In[2]:
+
+
+dump(lr, data_path / "lr_model.joblib")
+dump(testdf, data_path / "testdf.joblib")
```
Comment (gwaybio) [inline 1.train_models/log_reg_plate_1_cp_fs_data/nbconverted/train_model.py]: what is `testdf`? If it was for initial prototyping/testing, please consider deleting

File: 1.train_models/log_reg_plate_2_cp_fs_data/nbconverted/train_model.py [hunk 5]
```diff
@@ -0,0 +1,237 @@
+#!/usr/bin/env python
```
Comment (gwaybio) [inline 1.train_models/log_reg_plate_2_cp_fs_data/nbconverted/train_model.py]: I see a lot of redundant code here (redundant with the files in plate 1).

Remember [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) is best practices! Is there a way to refactor out functions in this code to help? (Sometimes DRY can be violated, in some circumstances)

File: 2.evaluate_models/log_reg_plate_2_cp_fs_data/nbconverted/evaluate_model.py [hunk 6]
```diff
@@ -0,0 +1,164 @@
+#!/usr/bin/env python
```
Comment (gwaybio) [inline 2.evaluate_models/log_reg_plate_2_cp_fs_data/nbconverted/evaluate_model.py]: Similar comment about some redundancy.

**Review by gwaybio — APPROVED** [review]
This looks great @MattsonCam - very good documentation and clear code.

I made some discussion comments, which probably should be addressed prior to merging.

Here is a summary:
1. Please make sure your notebooks are executed sequentially. (They are currently not executed)
2. Please consider DRY principles when writing code. Reducing redundency by factoring out functions is a good long-term gardening strategy.
3. Please consider using predefined functionality (e.g., `cross_validate` in sklearn) wherever possible to avoid reinventing wheels.

You may also want to consider using `precommit` `isort` to help organize import statements: https://pycqa.github.io/isort/docs/configuration/pre-commit.html

No need for me to re-review prior to you merging, but please do consider making the gardening improvements :)

File: 0.data_analysis/plate_projection_cp_fs_data/nbconverted/plate_1_2_projections.py [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+
+# ## Find the root directory of the repo regardless of repo location on system
+
+# In[ ]:
+
+
+# Get the current working directory
+cwd = Path.cwd()
+
+if (cwd / ".git").is_dir():
```
Comment (MikeLippincott) [inline 0.data_analysis/plate_projection_cp_fs_data/nbconverted/plate_1_2_projections.py:27]: Clever, I never thought of this and might start using this.

File: 0.data_analysis/plate_projection_cp_fs_data/nbconverted/plate_1_2_projections.py [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+# Create a dataframe with Null data
+platesnull = platescomb.loc[~platescomb["labels"].str.contains("WT")]
+
+
+# ## Visualization
+
+# In[1]:
+
+
+save_args = {"fname": out_path / "pca_plates_1_2"}
```
Comment (MikeLippincott) [inline 0.data_analysis/plate_projection_cp_fs_data/nbconverted/plate_1_2_projections.py:127]: Consider adding documentation on what function is being called. I really like the `**arg` implementation though.

**Review by MikeLippincott — COMMENTED** [review]
LGTM, great models. Most comments are regarding documentation of the helper function when calling. 

One thing to consider is that .pngs could have more descriptive file names so you know which file is which if you want to send them/export them for presentation. As in this case, they will be outside of their folders.

File: 0.data_analysis/plate_projection_cp_fs_data/nbconverted/plate_1_2_projections.py (hunk 1 above)
Comment (MattsonCam) [reply 0.data_analysis/plate_projection_cp_fs_data/nbconverted/plate_1_2_projections.py:8]: Ahh, that's good to know. I changed the sequence I am executing code.

File: 1.train_models/log_reg_plate_1_cp_fs_data/nbconverted/train_model.py (hunk 2 above)
Comment (MattsonCam) [reply 1.train_models/log_reg_plate_1_cp_fs_data/nbconverted/train_model.py]: The val_frac and the test_frac aren't necessary for this analysis since I'm using kfold. This is fixed now.

File: 1.train_models/log_reg_plate_1_cp_fs_data/nbconverted/train_model.py (hunk 3 above)
Comment (MattsonCam) [reply 1.train_models/log_reg_plate_1_cp_fs_data/nbconverted/train_model.py]: That's good to know, I suspected something like that existed, I've always just used KFold, but this might be a simpler approach.

I'll also use average_precision, it might matter, something that could also be explored.

File: 1.train_models/log_reg_plate_1_cp_fs_data/nbconverted/train_model.py (hunk 4 above)
Comment (MattsonCam) [reply 1.train_models/log_reg_plate_1_cp_fs_data/nbconverted/train_model.py]: testdf is for 2.evaluate_models

Comment (MattsonCam) [reply 1.train_models/log_reg_plate_1_cp_fs_data/nbconverted/train_model.py]: I was considering grouping the evaluation and the training parts of the models together, but I decided to keep them separate to follow the structure of another analysis repo. Maybe grouping them would be better?

File: 1.train_models/log_reg_plate_2_cp_fs_data/nbconverted/train_model.py (hunk 5 above)
Comment (MattsonCam) [reply 1.train_models/log_reg_plate_2_cp_fs_data/nbconverted/train_model.py]: Ya, I was thinking about that. I was trying to make each model and analysis as modular as possible.

In some cases the data wrangling may be slightly different for each dataset, but now that you mention it, I think it's a good idea to break this structure, in this case, and refactor with some functions for plates 1 and 2.

File: 2.evaluate_models/log_reg_plate_2_cp_fs_data/nbconverted/evaluate_model.py (hunk 6 above)
Comment (MattsonCam) [reply 2.evaluate_models/log_reg_plate_2_cp_fs_data/nbconverted/evaluate_model.py]: Agreed, I will apply the suggestion mentioned above.

File: 1.train_models/log_reg_plate_1_cp_fs_data/nbconverted/train_model.py (hunk 4 above)
Comment (gwaybio) [reply 1.train_models/log_reg_plate_1_cp_fs_data/nbconverted/train_model.py]: I think separating is better :)

also, if `testdf` is used downstream, please use a more descriptive file name. Also, is it a pandas dataframe? If so, a compressed csv is probably better than saving it as a joblib file.

Comment (MattsonCam) [reply 1.train_models/log_reg_plate_1_cp_fs_data/nbconverted/train_model.py]: Good point, I'll change that

**Comment by MattsonCam** [comment]
Merging now, thanks both of you!
